eliminate one of the data races on access to node data

seebs opened the pull request:
This particular data race hits us a lot in CI testing. The
underlying issue is that access to the contents of m.nodeMap
is controlled by a mutex, but rawSendMsgPacket is using the
mutex only to grab the pointer -- then accessing the contents
of the pointer without the mutex held. This means a data race
if something else grabs the mutex and modifies the contents
of the pointer as well.

In this particular case, access is read-only, so we copy the
pointed-to thing into a local. Less efficient, but also less
of a data race.

(Probably)
Fixes #113, #134 
(Well, not all of #134, but this was at least one of them.)

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/memberlist?pullRequest=179)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/memberlist?pullRequest=179)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/memberlist?pullRequest=179).</sub>

JensRantil commented on net.go:
```diff
@@ -678,10 +678,15 @@ func (m *Memberlist) rawSendMsgPacket(addr string, node *Node, msg []byte) error
+			node = &nodeCopy
 		}
+		m.nodeLock.RUnlock()
+
```
```suggestion
```

seebs replied:
yeah, probably. the extra blank line wasn't intentional.

jaffee commented:
Ping on this - the races are real and hit us occasionally. Thanks!

seebs commented:
pushed update to drop the spare newline. my branch should now be on 7e5a340e.

seebs commented:
So I got hit by one of them again, and studied the code, and realized that the branch I fixed wasn't probably the relevant branch -- it was in the case where `rawSendMsgPacket` *is* given a node pointer. On further study, it turns out there's a general pattern of using `kRandomNodes` to collect some nodes, then operate on them -- but while `kRandomNodes` is usually called with the node mutex, the resulting operations aren't.

Modifying `kRandomNodes` to copy the nodes it is returning seems to help. If I run the test that was hitting this 10 times without this patch, I "probably" get a data race. If I run it 100 times with, I don't see any races.

```
Write at 0x00c0005b7249 by goroutine 441:
  github.com/hashicorp/memberlist.(*Memberlist).aliveNode()
      /go/pkg/mod/github.com/pilosa/memberlist@v0.1.4-0.20190408132233-ff8741fd3108/state.go:999 +0xf16
  github.com/hashicorp/memberlist.(*Memberlist).mergeState()
      /go/pkg/mod/github.com/pilosa/memberlist@v0.1.4-0.20190408132233-ff8741fd3108/state.go:1187 +0x3b2
  github.com/hashicorp/memberlist.(*Memberlist).mergeRemoteState()
      /go/pkg/mod/github.com/pilosa/memberlist@v0.1.4-0.20190408132233-ff8741fd3108/net.go:1071 +0x642
  github.com/hashicorp/memberlist.(*Memberlist).handleConn()
      /go/pkg/mod/github.com/pilosa/memberlist@v0.1.4-0.20190408132233-ff8741fd3108/net.go:264 +0xa36

Previous read at 0x00c0005b7249 by goroutine 439:
  github.com/hashicorp/memberlist.(*Memberlist).rawSendMsgPacket()
      /go/pkg/mod/github.com/pilosa/memberlist@v0.1.4-0.20190408132233-ff8741fd3108/net.go:693 +0x8e3
  github.com/hashicorp/memberlist.(*Memberlist).gossip()
      /go/pkg/mod/github.com/pilosa/memberlist@v0.1.4-0.20190408132233-ff8741fd3108/state.go:534 +0x650
  github.com/hashicorp/memberlist.(*Memberlist).gossip-fm()
      /go/pkg/mod/github.com/pilosa/memberlist@v0.1.4-0.20190408132233-ff8741fd3108/state.go:489 +0x41
  github.com/hashicorp/memberlist.(*Memberlist).triggerFunc()
      /go/pkg/mod/github.com/pilosa/memberlist@v0.1.4-0.20190408132233-ff8741fd3108/state.go:135 +0x179
```

This is a sample of the trace, and the underlying issue is that `gossip` is grabbing a list of nodes and calling `rawSendMsgPacket`, which checks the state of the node it's invoked on, while `aliveNode` may be modifying that node's state. With the patch, `gossip` grabs the old node states, and operates on them, and if updates come in it misses them, but it's no longer a data race. (The other alternative might be to, say, request a lock again before evaluating the contents of the node, but I assume that if it were practical to run the entire operation with a read lock held, it would already be being done?)

seebs commented:
Oh, interesting. That test failure makes me think that the code *does* want to get those updates... In which case I'm not at all sure what a correct fix would look like.

banks commented:
Hey thanks for the PR @seebs there are several races caught by the detector here which is certainly not ideal. One issue with fixing them though is validating the performance hit caused e.g. by copying lots of stuff.

For context we know of users who run Consul clusters with over 10k nodes in one Gossip pool. For them any operation like PushPull that has to operate in O(n) over the number of nodes is already problematic so the performance impact of copying all the node states every time we pick a new node to probe (at least every second) would likely be prohibitive in that environment.

Do we know if anyone has observed actual incorrect behaviour due to these races or is it "just" being picked up by the race detector. I say "just" because I still think it's important to not have data races at all even if they don't seem to cause problems but it's important in deciding how to trade off against performance and time spent finding a solution that appeases the race detector (and is correct in operation) without impacting performance.

All that said, I don't _know_ this would be an issue to merge as-is but I suspect it would so without some in-depth performance analysis (by us or anyone else who is inclined to do it) it's risky for us! Testing these things at scale is pretty time consuming as you can imagine. A microbenchmark of the code paths changed might at least give some basic data we can use to get an idea of the potential issue.

If you have any more insights on that or are interested in finding possible solutions that might impact performance less we're interested to hear. We hope to get to investigating these races more at some point however as always it's a prioritisation problem - lots of work to do vs. not clear if it's actually impacting users in production (outside of CI complaining which is also a legit problem but a different one!).

seebs commented:
Yeah, the performance hit is definitely a problem with this fix. I'm not sure what alternatives exist. We have a similar thing in our code base where I implemented it as copy-on-write; when things get returned by something like kRandomNodes, they get a "frozen" flag set; anything that wants to change them, thereafter, will actually replace them with a new object. So the object gets copied if and only if it's being written after being copied out by something that expects it to be read outside the lock. This design is working quite nicely for us, so far.

I don't know about actual incorrect behavior due to the races, but anecdotally, I think our incidence rate of Weird Gossip Failures We Can't Directly Explain went down after we changed this, even though the motivation for changing it was just a ~1-2% failure rate of any and all CI tests using `-race`.

banks commented:
Interesting, thanks for sharing your experience. Just to be sure - you
deployed your patch here in production? What kind of size is the cluster if
you can share that? Any noticeable increase in CPU from your processes?

On Thu, May 23, 2019 at 3:18 PM seebs <notifications@github.com> wrote:

> Yeah, the performance hit is definitely a problem with this fix. I'm not
> sure what alternatives exist. We have a similar thing in our code base
> where I implemented it as copy-on-write; when things get returned by
> something like kRandomNodes, they get a "frozen" flag set; anything that
> wants to change them, thereafter, will actually replace them with a new
> object. So the object gets copied if and only if it's being written after
> being copied out by something that expects it to be read outside the lock.
> This design is working quite nicely for us, so far.
>
> I don't know about actual incorrect behavior due to the races, but
> anecdotally, I think our incidence rate of Weird Gossip Failures We Can't
> Directly Explain went down after we changed this, even though the
> motivation for changing it was just a ~1-2% failure rate of any and all CI
> tests using -race.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hashicorp/memberlist/pull/179?email_source=notifications&email_token=AAA5QU7XQBKD6V3ZOGAHBKDPW2RTHA5CNFSM4GSFSIJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWCL2RY#issuecomment-495238471>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA5QU353XY2XWZ2J6345B3PW2RTHANCNFSM4GSFSIJA>
> .
>

seebs commented:
I haven't looked at "real" production systems, and I've only been doing small clusters. It's probably not-zero, but it's not large enough to be obvious... But it's pretty obvious to me that it's going to be a potential problem. For *my* purposes, "fix the problem so I stop getting CI failures" is a good choice. For larger clusters, I'd probably want to actually implement something like copy-on-write.

The basic idea is: things like kRandomNodes, which will return nodeStates to things outside a lock, set a "frozen" flag. Anything which would be writing to the data that gets read outside the lock checks for that flag, and if it's set, then copies that specific node, modifies the new one, and stores the new pointer in whatever list it's looking at. This works in our system, because at any given point, usually only one thing has a pointer to a given object of the type in question at once, unless it's a transient "copy" that's being used to continue accessing the data outside a lock.

For read-heavy workloads, things get marked frozen and stay that way, for write-heavy workloads, they don't get frozen, we only copy them a lot when there's both reads and writes. But with a large cluster, the assumption would be that, while nodes change state fairly often, they don't *all* change state all that often, so while there'd still be a lot of new nodeState objects coming into existence, it wouldn't be "all k of them, unconditionally, every time we want kRandomNodes".

The other concern I've got is related to the CI problem: Even if the data races aren't actually resulting in corruption (which I'd be slightly skeptical of; it seems like it's possible to get inconsistent reads in a way that would make nonsensical choices happen), they're resulting in not running with race detector enabled, meaning *other* data races are *also* going undetected. Keeping stuff race-free lets us use potentially important tools.

banks commented:
Thanks.

I totally agree getting it race free is important even just for CI but also
because it's dangerous to leave it racey!

On Thu, May 23, 2019 at 4:05 PM seebs <notifications@github.com> wrote:

> I haven't looked at "real" production systems, and I've only been doing
> small clusters. It's probably not-zero, but it's not large enough to be
> obvious... But it's pretty obvious to me that it's going to be a potential
> problem. For *my* purposes, "fix the problem so I stop getting CI
> failures" is a good choice. For larger clusters, I'd probably want to
> actually implement something like copy-on-write.
>
> The basic idea is: things like kRandomNodes, which will return nodeStates
> to things outside a lock, set a "frozen" flag. Anything which would be
> writing to the data that gets read outside the lock checks for that flag,
> and if it's set, then copies that specific node, modifies the new one, and
> stores the new pointer in whatever list it's looking at. This works in our
> system, because at any given point, usually only one thing has a pointer to
> a given object of the type in question at once, unless it's a transient
> "copy" that's being used to continue accessing the data outside a lock.
>
> For read-heavy workloads, things get marked frozen and stay that way, for
> write-heavy workloads, they don't get frozen, we only copy them a lot when
> there's both reads and writes. But with a large cluster, the assumption
> would be that, while nodes change state fairly often, they don't *all*
> change state all that often, so while there'd still be a lot of new
> nodeState objects coming into existence, it wouldn't be "all k of them,
> unconditionally, every time we want kRandomNodes".
>
> The other concern I've got is related to the CI problem: Even if the data
> races aren't actually resulting in corruption (which I'd be slightly
> skeptical of; it seems like it's possible to get inconsistent reads in a
> way that would make nonsensical choices happen), they're resulting in not
> running with race detector enabled, meaning *other* data races are *also*
> going undetected. Keeping stuff race-free lets us use potentially important
> tools.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hashicorp/memberlist/pull/179?email_source=notifications&email_token=AAA5QU5G3DTFMXQS324DAMTPW2XFNA5CNFSM4GSFSIJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWCQWMA#issuecomment-495258416>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA5QU6WOECD2AZ3EEZYCBTPW2XFNANCNFSM4GSFSIJA>
> .
>

seebs commented:
So, long story short, I was talking about a race condition thing elsewhere, and I got to thinking.

`sync/atomic` is safe against data races, sort of. It's not safe in that it doesn't provide guarantees that you don't get incoherent data if you read multiple values, but it's safe in that it's not actual full on data races that the race detector cares about.

But it turns out, there's not that many accesses to these nodes, and in practice, I don't think I ever saw one of these races hit where the race didn't involve the read in rawSendMsgPacket. What's that reading from the node? So far as I can tell, the only thing it ever reads is `node.PMax`.

In general, the pointers themselves don't really become invalidated; they're added to m.nodeMap only if no previous state pointer existed, otherwise the state gets updated in-place. And everything is then modifications throught that state pointer.

So... It'd probably be safe to do these with atomic ops. Except now that I look, the fields in question are `uint8` and there's no 8-bit atomic ops in sync/atomic, just 32/64/pointer.

But I've also realized that it's quite possible that those values never actually change in real-world situations. After all, it's quite possible that a given node continues supporting the same protocol versions for its entire functional lifespan, or at least that updates to memberlist are rare. And parallel reads aren't data races. So if those all became

```
if state.PMax != a.Vsn[1] {
    state.PMax = a.Vsn[1]
}
```

it'd probably never trigger again. Or at least, hardly ever.

banks commented:
It seems like the simplest solution that avoids copies would be to change all call sites that update nodes from this map to do so via a central method that takes the lock. I've not looked through the code to see how many places that is but from your discussion @seebs it seems to be only a few right? Is there some reason that is going to be a difficult fix? I don't think lock contention is too much of a problem given that all the reads are already contending on the same lock.

seebs commented:
I think everything updating the nodes already has a lock. The race condition is against reads of lists returned from kRandomNodes, I think? And those lists are around for a while after they're obtained; the problem is that the list is obtained under the lock, but then the pointers are dereferenced outside it, which is why my first patch was just to make copies. (Slow but almost certainly correct.)

banks commented:
Ah got it. I'm not dead against copying provided we are confident the
performance cost isn't significant when you have large clusters.

The rate of updates remains reasonably constant (push pull increases with a
larger cluster but logarithmically rather than linearly but regular gossip
rate doesn't). The killer is the linear growth in number of nodes in the
list. If we could simulate/test with ~10k nodes and show CPU about the same
with copying I'd be fine merging it.

Copy on write would also be a valid solution which might not be too hard to
build depending on how many different code paths perform writes. If they
are all taking lock already then they could just copy and update and
replace in the map under lock and all would be happy presumably - the
readers holding pointers to the old struct would still be using a stale
value but that's OK and already true, but they wouldn't be dereferencing a
pointer to a struct that might later be modified...

Apologies this is all pretty hand wavy as I have not had time to dive into
the code to remind myself of the details!


On Wed, Sep 18, 2019 at 4:34 PM seebs <notifications@github.com> wrote:

> I think everything updating the nodes already has a lock. The race
> condition is against reads of lists returned from kRandomNodes, I think?
> And those lists are around for a while after they're obtained; the problem
> is that the list is obtained under the lock, but then the pointers are
> dereferenced outside it, which is why my first patch was just to make
> copies. (Slow but almost certainly correct.)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hashicorp/memberlist/pull/179?email_source=notifications&email_token=AAA5QU4RQEFIRPKO5GFZBEDQKJDBRA5CNFSM4GSFSIJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7APUVA#issuecomment-532740692>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA5QU6RHM6HIV27XSM27A3QKJDBRANCNFSM4GSFSIJA>
> .
>
